connector/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryClient.java

```diff
@@ -168,32 +171,57 @@ String fullTableName(TableId tableId) {
     return format("%s.%s.%s", tableId.getProject(), tableId.getDataset(), tableId.getTable());
   }
 
-  public long calculateTableSize(TableId tableId, Optional<String> filter) {
+  public TableStatistics calculateTableSize(TableId tableId, Optional<String> filter) {
     return calculateTableSize(getTable(tableId), filter);
   }
 
-  public long calculateTableSize(TableInfo tableInfo, Optional<String> filter) {
-    try {
-      TableDefinition.Type type = tableInfo.getDefinition().getType();
-      if (type == TableDefinition.Type.TABLE && !filter.isPresent()) {
-        return tableInfo.getNumRows().longValue();
-      } else if (type == TableDefinition.Type.VIEW
-          || type == TableDefinition.Type.MATERIALIZED_VIEW
-          || (type == TableDefinition.Type.TABLE && filter.isPresent())) {
-        // run a query
-        String table = fullTableName(tableInfo.getTableId());
-        String sql = format("SELECT COUNT(*) from `%s` WHERE %s", table, filter.get());
-        TableResult result = bigQuery.query(QueryJobConfiguration.of(sql));
-        return result.iterateAll().iterator().next().get(0).getLongValue();
-      } else {
-        throw new IllegalArgumentException(
-            format(
-                "Unsupported table type %s for table %s",
-                type, fullTableName(tableInfo.getTableId())));
-      }
-    } catch (InterruptedException e) {
-      throw new BigQueryConnectorException(
-          "Querying table size was interrupted on the client side", e);
+  public TableStatistics calculateTableSize(TableInfo tableInfo, Optional<String> filter) {
+    TableDefinition.Type type = tableInfo.getDefinition().getType();
+    if (type == TableDefinition.Type.TABLE) {
+      // if this is a table, there's no need to query the total number of rows
+      OptionalLong numberOfFilteredRows =
+          toOptionalLong(
+              filter
+                  .map(f -> countWithFilter(tableInfo.getTableId(), f))
+                  .map(row -> row.get(0).getLongValue()));
+      return new TableStatistics(tableInfo.getNumRows().longValue(), numberOfFilteredRows);
+    } else if (type == TableDefinition.Type.VIEW
+        || type == TableDefinition.Type.MATERIALIZED_VIEW) {
+      // For views, we need to calculate the number of rows
+      String table = fullTableName(tableInfo.getTableId());
+      String sql =
+          filter
+              .map(
+                  f ->
+                      format(
+                          "SELECT (SELECT COUNT(*) from `%s`), (SELECT COUNT(*) from `%s` WHERE %s)",
```

Comment:
did you compare the plan for this with something like "Select count(*), sum(filter) from table?

Reply:
+1. I think this query should have two input stages (one with pushdown and one without). Would be interesting to compare plans/perf.

Reply:
Replaced with `format("SELECT COUNT(*), COUNTIF(%s) from `%s`", f, table)`